src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -20,26 +22,95 @@ namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage
+            where TVirtualDocumentSnapshot : VirtualDocumentSnapshot
+        {
         }
 
+        public override Task<SynchronizedResult<TVirtualDocumentSnapshot>> TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(
+            int requiredHostDocumentVersion,
+            Uri hostDocumentUri,
+            Uri virtualDocumentUri,
+            CancellationToken cancellationToken)
+            where TVirtualDocumentSnapshot : class
+            => TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(
+                requiredHostDocumentVersion,
+                hostDocumentUri,
+                virtualDocumentUri,
+                rejectOnNewerParallelRequest: true,
+                cancellationToken);
+
+        public override async Task<SynchronizedResult<TVirtualDocumentSnapshot>> TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(
+            int requiredHostDocumentVersion,
+            Uri hostDocumentUri,
+            Uri virtualDocumentUri,
+            bool rejectOnNewerParallelRequest,
+            CancellationToken cancellationToken)
+            where TVirtualDocumentSnapshot : class
+        {
+            if (hostDocumentUri is null)
+            {
+                throw new ArgumentNullException(nameof(hostDocumentUri));
+            }
+
+            Task<bool> onSynchronizedTask;
+            lock (_documentContextLock)
+            {
+                if (!_virtualDocumentContexts.TryGetValue(virtualDocumentUri, out var documentContext))
+                {
+                    var removedDocumentSnapshot = GetVirtualDocumentSnapshot<TVirtualDocumentSnapshot>(hostDocumentUri);
+                    // Document was deleted/removed in mid-synchronization
+                    return new SynchronizedResult<TVirtualDocumentSnapshot>(false, removedDocumentSnapshot);
+                }
+
+                if (requiredHostDocumentVersion == documentContext.SeenHostDocumentVersion)
+                {
+                    var synchronziedDocumentSnapshot = GetVirtualDocumentSnapshot<TVirtualDocumentSnapshot>(hostDocumentUri);
+                    // Already synchronized
+                    return new SynchronizedResult<TVirtualDocumentSnapshot>(true, synchronziedDocumentSnapshot);
+                }
+
+                // Currently tracked synchronizing context is not sufficient, need to update a new one.
+                onSynchronizedTask = documentContext.GetSynchronizationTaskAsync(requiredHostDocumentVersion, rejectOnNewerParallelRequest, cancellationToken);
+            }
+
+            var onSynchronizedResult = await onSynchronizedTask;
+
+            var virtualDocumentSnapshot = GetVirtualDocumentSnapshot<TVirtualDocumentSnapshot>(hostDocumentUri);
+
+            return new SynchronizedResult<TVirtualDocumentSnapshot>(onSynchronizedResult, virtualDocumentSnapshot);
+        }
+
+        [Obsolete]
```

Comment:
Obsoleting these instead of deleting because I'm told WebTools relies on this class? Or was that just on the requestInvoker? Would prefer to delete them so if it's just the invoker let me know and I'll kill these.

Reply:
Web Tools definitely uses our ContainedLanguage library